src/analyses/mallocWrapperAnalysis.ml

```diff
@@ -57,28 +57,29 @@ struct
 
 
   (* refs to reassign unmarshaled in init *)
-  let heap_hash = ref (Hashtbl.create 113)
-  let heap_vars = ref (Hashtbl.create 113)
+  module NH = Hashtbl.Make (Node)
+  module VH = Hashtbl.Make (CilType.Varinfo)
+  let heap_hash = ref (NH.create 113)
+  let heap_vars = ref (VH.create 113)
 
   type marshal = {
-    heap_hash: (string, varinfo) Hashtbl.t;
-    heap_vars: (int, unit) Hashtbl.t;
+    heap_hash: varinfo NH.t;
+    heap_vars: unit VH.t;
   }
 
   let get_heap_var node =
     (* Use existing varinfo instead of allocating a duplicate,
        which would be equal by determinism of create_var though. *)
     (* TODO: is this poor man's hashconsing? *)
-    let nodeId = match node with
-      | Node.Statement s -> "sid:" ^ (string_of_int s.sid)
-      | Function f -> "vid:" ^ (string_of_int f.svar.vid)
-      | _ -> raise (Failure "A function entry node can never be the node after a malloc") in
-    try Hashtbl.find !heap_hash nodeId
+    try NH.find !heap_hash node
     with Not_found ->
-      let name = "(alloc@" ^ nodeId ^ ")" in
+      let name = match node with
+        | Node.Statement s -> "(alloc@sid:" ^ (string_of_int s.sid) ^ ")"
+        | Function f -> "(alloc@vid:" ^ (string_of_int f.svar.vid) ^ ")"
+        | _ -> raise (Failure "A function entry node can never be the node after a malloc") in
```

Comment:
Can a function return node actually `malloc` something?

Reply:
The node that is used when creating a heap variable, is the one following the call to `malloc`. The return edges will probably never contain a `malloc`(-Wrapper) call since CIL seems to introduce a temporary variable for function calls in the return statement. In that case, one would actually not need to consider return nodes when creating heap variables.

Reply:
Unless we know of an example where the `Function` case is necessary here, I think we should remove it and make it also raise an exception (you can use `failwith` directly). That would simplify the names a bit to always be statement IDs.